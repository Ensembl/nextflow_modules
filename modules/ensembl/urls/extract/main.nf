process URLS_EXTRACT {
    tag "$meta.id"
    label 'process_low'

    input:
    tuple val(meta), val(url_string)

    output:
    tuple val(meta), env(user), env(host), env(port), emit: connection_info
    tuple val(meta), env(password), emit: optional_password, optional: true
    tuple val(meta), env(database), emit: optional_database, optional: true
    path "versions.yml", emit: versions

    script:

    """
    # Extract and parse the MySQL URL
    url_string="${url_string}"

    # Use regex to extract all components
    if [[ "\$url_string" =~ ^mysql://([^:@]+)(:([^@]*))?@([^:]+):([0-9]+)(/([^/]*)?)?\$ ]]; then
        user="\${BASH_REMATCH[1]}"
        password="\${BASH_REMATCH[3]}"
        host="\${BASH_REMATCH[4]}"
        port="\${BASH_REMATCH[5]}"
        database="\${BASH_REMATCH[7]}"

        # Handle missing password or database
        [ -z "\$password" ] && password=""
        [ -z "\$database" ] && database=""
        export user host port password database
    else
        echo "Invalid MySQL URL format: \$url_string" >&2
        exit 1
    fi

    echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
    """

    stub:
    """
    host = host.com
    user = user
    port = 3306

    echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
    """
}