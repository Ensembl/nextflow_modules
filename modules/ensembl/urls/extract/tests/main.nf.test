nextflow_process {

    name "Test Process URLS_EXTRACT"
    script "../main.nf"
    process "URLS_EXTRACT"

    tag "modules"
    tag "modules_ensembl"
    tag "urls"
    tag "urls/extract"

    test("Should extract connection info from URL with all components") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test1' ], 
                    "mysql://user:password@host.com:3306/database"
                ]
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }
    }

    test("Should extract connection info from URL without password") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test2' ],
                    "mysql://user@host.com:3306/database"
                ]
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }
    }

    test("Should extract connection info from URL without database") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test3' ],
                    "mysql://user:password@host.com:3306"
                ]
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }
    }

    test("Should fail with invalid URL format") {
        when {
            process {
                """
                input[0] = [
                    [ id:'test4' ],
                    "invalid_url"
                ]
                """
            }
        }

        then {
            assert process.failed
            assert snapshot(process.stderr).match()
        }
    }
}

