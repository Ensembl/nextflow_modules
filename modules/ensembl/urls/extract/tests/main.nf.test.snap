{
    "optional_database": {
        "content": [
            [
                [
                    {
                        "id": "test1"
                    },
                    "database"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:46.911446"
    },
    "optional_password": {
        "content": [
            [
                [
                    {
                        "id": "test1"
                    },
                    "password"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:46.909742"
    },
    "Should extract connection info from URL without password": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test2"
                        },
                        "user",
                        "host.com",
                        "3306"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test2"
                        },
                        ""
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test2"
                        },
                        "database"
                    ]
                ],
                "3": [
                    "versions.yml:md5,844210d41d31414da2041f13946f845c"
                ],
                "connection_info": [
                    [
                        {
                            "id": "test2"
                        },
                        "user",
                        "host.com",
                        "3306"
                    ]
                ],
                "optional_database": [
                    [
                        {
                            "id": "test2"
                        },
                        "database"
                    ]
                ],
                "optional_password": [
                    [
                        {
                            "id": "test2"
                        },
                        ""
                    ]
                ],
                "versions": [
                    "versions.yml:md5,844210d41d31414da2041f13946f845c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T14:47:16.366839"
    },
    "Should extract connection info from URL without database": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test3"
                        },
                        "user",
                        "host.com",
                        "3306"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test3"
                        },
                        "password"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test3"
                        },
                        ""
                    ]
                ],
                "3": [
                    "versions.yml:md5,844210d41d31414da2041f13946f845c"
                ],
                "connection_info": [
                    [
                        {
                            "id": "test3"
                        },
                        "user",
                        "host.com",
                        "3306"
                    ]
                ],
                "optional_database": [
                    [
                        {
                            "id": "test3"
                        },
                        ""
                    ]
                ],
                "optional_password": [
                    [
                        {
                            "id": "test3"
                        },
                        "password"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,844210d41d31414da2041f13946f845c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T14:47:18.384323"
    },
    "optional_password_no_database": {
        "content": [
            [
                [
                    {
                        "id": "test3"
                    },
                    "password"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:51.085612"
    },
    "optional_password_no_password": {
        "content": [
            [
                [
                    {
                        "id": "test2"
                    },
                    "null"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:49.065466"
    },
    "connection_info": {
        "content": [
            [
                [
                    {
                        "id": "test1"
                    },
                    "user",
                    "host.com",
                    "3306"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:46.9039"
    },
    "Should extract connection info from URL with all components": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1"
                        },
                        "user",
                        "host.com",
                        "3306"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test1"
                        },
                        "password"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test1"
                        },
                        "database"
                    ]
                ],
                "3": [
                    "versions.yml:md5,844210d41d31414da2041f13946f845c"
                ],
                "connection_info": [
                    [
                        {
                            "id": "test1"
                        },
                        "user",
                        "host.com",
                        "3306"
                    ]
                ],
                "optional_database": [
                    [
                        {
                            "id": "test1"
                        },
                        "database"
                    ]
                ],
                "optional_password": [
                    [
                        {
                            "id": "test1"
                        },
                        "password"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,844210d41d31414da2041f13946f845c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T14:47:14.131667"
    },
    "optional_database_no_database": {
        "content": [
            [
                [
                    {
                        "id": "test3"
                    },
                    "null"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:51.087938"
    },
    "Should fail with invalid URL format": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T14:47:20.326552"
    },
    "optional_database_no_password": {
        "content": [
            [
                [
                    {
                        "id": "test2"
                    },
                    "database"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:49.067826"
    },
    "connection_info_no_password": {
        "content": [
            [
                [
                    {
                        "id": "test2"
                    },
                    "user",
                    "host.com",
                    "3306"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:49.062974"
    },
    "connection_info_no_database": {
        "content": [
            [
                [
                    {
                        "id": "test3"
                    },
                    "user",
                    "host.com",
                    "3306"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-26T12:22:51.083091"
    }
}